lisp-modules: replace with an implementation based on github:uthar/nix-cl

Uthar opened the pull request:
###### Description of changes

Replace lisp-modules with implementation based on https://github.com/uthar/nix-cl.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

1. Added Lisp section to Nixpkgs manual
2. Replaced `lisp-modules` with a new API. The interface consists of : 
- `sbcl.pkgs`
- `sbcl.withPackages` 
- `sbcl.buildASDFSystem`
- `sbcl.pkgs.alexandria.overrideLispAttrs (oldAttrs: { ... })` 
- `sbcl.withOverrides (self: super: { ... })`
- ~`sbcl.override { packageOverrides = { ... }; }`~
- ~`sbcl.override { spec = { asdf = ...; pkg = ...; program = ...; flags = ...; faslExt = ...;  }; }`~
- also for `abcl`, `ccl`, `clasp-common-lisp`, `clisp`, `cmucl_binary`, `ecl`, `mkcl`
3. Added `recurseIntoAttrs` for just `sbclPackages` for now (maybe extend with other CLs in the future?)
4. ~Removed~ Deprecated `lispPackages` and `lispPackages_new`
5. Updated `clasp-common-lisp` to release `2.2.0`

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Uthar committed:
lisp-modules: remove both current implementations

In preparation for including another API based on github:uthar/nix-cl

This commit will surely temporarily break some packages, but at least it will
make reviewing such big changes.

Uthar committed:
lisp-modules: Add manual section

Uthar committed:
clasp: remove

Uthar committed:
clasp: init at 2.2.0

Uthar committed:
asdf_3_3: init at 3.3.6

Uthar committed:
lisp-modules: init

From https://github.com/Uthar/nix-cl
Commit ebeb76057abf01029067f1ebe0ae6fae90848bfe

Uthar committed:
lisp-modules: wrap lisps

Uthar commented:
Left to discuss:

- [ ] Flags is an array of strings
- [ ] patches = [] defaults to empty anyway
- [ ] checksummed vs latest

Uthar commented:
- [X] I have used conventional prefixes in commits
- [X] I have split the patch in to smaller commits
- [X] I have Rewritten documentation
- [X] I have changed v1.4 -> 1.4 in alexandria examples
- [X] I have removed occurences of fetchTarball
- [X] I have made it so packages.sqlite is removed before import
- [X] I have fixed the weird alexantdria.overrideLispAttrs example
- [X] I have Moved back ecl.pkg -> ecl in some places
- [X] I have Removed the unnecessary gitignore
- [X] I have removed extraneous setup-hook.nix
- [X] I have documented that the previous approach didn't trust the upstream QL data
- [X] I have switched to :nix-cl-user instead of :cl-user in builder

Uthar committed:
stumpwm: fix after major changes to lisp-modules

Uthar committed:
nyxt: fix after major changes to lisp-modules

7c6f434c commented on doc/languages-frameworks/lisp.section.md:
````diff
@@ -0,0 +1,282 @@
+functions in the `org.lispbuilds.nix` ASDF system. To run the script, one can
+execute `ql-import.lisp`:
+
+```
+$ nix-shell
+$ sbcl --script ql-import.lisp
+```
````
If it's non-interactive, a one-liner is more convenient…

Uthar replied:
OK 
https://github.com/NixOS/nixpkgs/compare/f7d59f31a8a5a7b34a17772cbbc0447fa73baa73..6f26e173cf901372eb22e88d0c1a786984eb75ad#diff-e863f334ae54fd0ec4369e5fbe4fb69e36a4f1bcc2c69968cc0e57566b2c21e1

Uthar committed:
lisp-modules: make spec.flags a list of strings

7c6f434c commented on doc/languages-frameworks/lisp.section.md:
```diff
@@ -0,0 +1,321 @@
+should be included in the parent package as an extra entry in the `systems`
+argument to the `build-asdf-system`/`buildASDFSystem` functions.
+
+The reason is that ASDF searches for a secondary system in the `.asd` of the
+parent package. Thus, having them separate would cause either one of them not to
+load cleanly, because one will contains FASLs of itself but not the other, and
+vice versa.
```
Of course, there are some bad cases like `cl-ppcre/unicode` → `cl-unicode` → `cl-ppcre` (if I remember correctly)… I think even if there is nothing encouraging to tell about such situations, maybe it is better to mention as something to look out for?

Uthar replied:
Are you thinking about circular dependencies?

What would you add to the manual?

7c6f434c replied:
Well, if we have no generic advice to offer, we could add after the example 
«
Note that sometimes the slashy systems might not only have more dependencies than the main one, but create a circular dependency between `.asd` files. Unfortunately, in this case an adhoc solution becomes necessary.
»

Uthar replied:
Added https://github.com/NixOS/nixpkgs/compare/457303158f1fb67dded7b6d89f75a122c911ac14..2f70da2da9d57291024e2989b9016f5f49924fb6

7c6f434c commented on pkgs/development/lisp-modules/COPYING:
```diff
@@ -0,0 +1,22 @@
+Copyright 2021-2023 Kasper Gałkowski
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
```
I am tempted to ask, if there is an upstream repo for this code, and the license itself is different from Nixpkgs, is it a good idea to just _package_ `nix-cl`.

If not, I think it is better to create a subdirectory for the generator and put the files there. Both from the point of view of not mixing the generator codebase with the output, and also to be able to separate the Nix-evaluation-time files from this licensing (presumably we will get to three or five people editing the output side pretty quickly)

Uthar replied:
I can remove this COPYING file, then Nixpkgs license will apply

Uthar replied:
The code for importing from Quicklisp is already in the separate `import` directory

7c6f434c replied:
Ah right, indeed the builder is another story.

I think as is the situation with the COPYING file is going towards implying false claims, so it needs fixing. I guess either packaging-from-upstream-repo or letting match Nixpkgs are perfect if you are OK with either, otherwise it's fine to isolate into subdirs so that COPYING would cover something clearly separated from things intended to be immediately edited by others.

Uthar replied:
I think I will just remove the COPYING here. 

But I'm not sure from your comment, do you think removing the COPYING file here will solve the problem?

7c6f434c replied:
I think so. Nixpkgs has various pieces of code contributed for inclusion into the Nixpkgs under the current license, you are the sole author so you can choose to submit for inclusion under MIT license. And Nixpkgs license does refer to contributors as people coauthoring Nixpkgs, instead of listing specific people by name. I guess the end user will have a choice of taking your code from BSD2 upstream or from MIT downstream.

I was just against «put COPYING where it seems to cover a file that will be edited by people not listed in closed-list-of-people COPYING», and also a bit against having BSD2-licensed Nix expressions in MIT-licensed Nixpkgs.

7c6f434c commented on pkgs/development/lisp-modules/import/main.lisp:
```diff
@@ -0,0 +1,42 @@
+  (format t "Dumped nix file to ~a~%"
+          (truename "imported.nix")))
+
+(defun main ()
```
Maybe add support for using the latest quicklisp release if a command-line flag is passed? 

If we already want `@var@` substitutions, we could at least control the version used via something Nix-visible. Or just keep the current quicklisp package, and use version from there.

Uthar replied:
> Maybe add support for using the latest quicklisp release if a command-line flag is passed?

This sounds better to me, But also in my opinion updating the file once in a while is not bad. But it's not too hard to implement, so I won't oppose.

Do you think it should be added now or in a future patch?

7c6f434c replied:
I think having a version pin that is not complete enough for fixed-output derivation, not auto-updatable, not command-line controllable, and not visible from Nix, oh and also it is in the middle of a source file with some meaningful code too — it is too much. If we update is in a source file, while not just keep the quicklisp package and take files from there? And then an auto-updater can be added some day.

Uthar replied:
OK, I made it so it always takes the latest Quicklisp version https://github.com/NixOS/nixpkgs/pull/218870/commits/57df1050133e4591ba2b0544e15f7a54462ac405

Uthar replied:
Do you think that solves the problem?

7c6f434c replied:
I think for practical purposes yes, there are some people with huge global ideas who will be unhappy about lack of the pinned options, but I am currently busy conspiciously not buying their bridge anyway.

7c6f434c commented on pkgs/development/lisp-modules/import/database/sqlite.lisp:
```diff
@@ -0,0 +1,176 @@
+   "mcclim-bezier"
+   ;; Missing dependency on c2ffi cffi extension
+   "hu.dwim.zlib"
+   ;; Missing libgvc.so native library
```
Well, it is in `graphviz`, will it work if we pass it?

Uthar replied:
Fixed in dcd555f3169b06c18536370a605093d91aa6829f

7c6f434c replied:
Thanks

Uthar committed:
sbclPackages.hu_dot_dwim_dot_graphviz: fix build

hu.dwim.graphviz didn't build because of lack of graphiz in nativeLibs

Uthar commented:
I should also document that the `imported.nix` file should not be edited by hand.

7c6f434c commented on pkgs/development/lisp-modules/import/nix.lisp:
```diff
@@ -0,0 +1,81 @@
+  (flet ((fix-special-chars (str)
+           (replace-regexes '("[+]$" "[+][/]" "[+]" "[.]" "[/]")
+                            '("_plus" "_plus/" "_plus_" "_dot_" "_slash_")
+                            str)))
```
I would also replace `_` with `_underscore_` for reversibility.

Uthar replied:
I did it locally.  It affected 363 names. Hmm. I'm not sure if this is good. It results in a name such as `cl-glfw-opengl-nv_underscore_vertex_underscore_buffer_underscore_unified_underscore_memory` which is very long. Do you think we can skip replacing `_` with `_underscore_`?

7c6f434c replied:
Hm, maybe `__` would be not too bad then?

Uthar replied:
OK, added [57563cf](https://github.com/NixOS/nixpkgs/pull/218870/commits/57563cf8f0126169809259124cf905c429820f22)

7c6f434c replied:
Thanks!

Uthar committed:
lisp-modules: make ql-import.lisp automatically pull the latest Quicklisp dist

Uthar committed:
sbclPackages.cl-gtk4: update to tip of trunk

Makes latest examples from upstream work

7c6f434c commented on pkgs/development/lisp-modules/ql.nix:
```diff
@@ -0,0 +1,209 @@
+      nativeLibs = [ pkgs.readline ];
+    });
+    md5 = super.md5.overrideLispAttrs (o: {
+      lispLibs = [ super.flexi-streams ];
```
What exactly is happening there that _Lisp_ deps need to be corrected?

Uthar replied:
Fixes build on non-SBCL

`md5` uses `sb-rotate-byte` on SBCL and uses `flexi-streams` on other implementations. I guess Quicklisp is built using SBCL and that's why the dependency is missing.

Uthar replied:
now there is (-: https://github.com/quicklisp/quicklisp-projects/issues/2266

7c6f434c replied:
Thanks

It's more complicated than you decribe: with ql:quickload it will still work because deps are just hints and QL hooks into ASDF… (which is why _completely_ broken dependencies could persist for mutiple QL releases in a row)

Uthar commented on pkgs/development/compilers/clasp/default.nix:
```diff
@@ -1,129 +1,116 @@
-}:
+{ pkgs, lib, fetchFromGitHub, llvmPackages_15
+# For packages
+, asdf_3_3
+, commonLispPackagesFor
+, lispWithPackages
+, build-asdf-system
+, spec ? { faslExt = "fasp"; program = "clasp"; flags = []; asdf = asdf_3_3; }
+, packageOverrides ? (self: super: {})
```
Hmm, I wasn't able to come up with such a solution. I didn't know how to compose multiple function arguments lists into one function.

7c6f434c replied:
You just need `passthru`, right? so something like `wrapLisp= p: p // (passthruFor p)` applied to the package as made by `mkDerivation` should work?

Uthar replied:
I read some code of `python-modules` and `lua-modules` and come up with something that looks cleaner (at least to me)

See commit [4f0e2a2](https://github.com/NixOS/nixpkgs/pull/218870/commits/4f0e2a2fb0aeb4f4abd7d5c58ba7fd4aa0b59364)

Uthar replied:
Hmm, maybe making `packageOverrides` an argument to `wrapLisp` would be even simpler ...

Uthar replied:
Alright, I added a new funciton `withOverrides` to `abcl`, `ecl` etc. https://github.com/NixOS/nixpkgs/pull/218870/commits/dae0dca5d1c9ad4150b91c55c5f4b54430b1413c

Let me know what you think

7c6f434c replied:
I think it's generally fine. And passthru stuff wouldn't be fragile to some of the overrides either, I think.

Uthar commented on pkgs/development/lisp-modules/ql.nix:
```diff
@@ -0,0 +1,209 @@
+
+  overrides = (self: super: {
+    cl_plus_ssl = super.cl_plus_ssl.overrideLispAttrs (o: {
+      nativeLibs = [ pkgs.openssl ];
```
Hmm, I run the test suite, looks OK:

```
NIX_PATH=nixpkgs=$(pwd) nix-shell -p 'sbcl.withPackages (ps: [ ps.cl_plus_ssl_dot_test ])' --run "sbcl --eval '(load (sb-ext:posix-getenv \"ASDF\"))' --eval '(asdf:load-system :cl+ssl.test)' --eval '(5am:run :cl+ssl)'"
```

7c6f434c replied:
Ah nice, update fixes compatibility! (hopefully)

Uthar committed:
lisp-modules: replace "_" in package names with "__" for reversibility

Uthar committed:
sbclPackages.math: fix build

There was a compile-time pathname into the home directory. It was patched to be
a function that works on run-time.

Uthar commented on pkgs/development/lisp-modules/import/database/sqlite.lisp:
```diff
@@ -0,0 +1,176 @@
+   "hu.dwim.logger"
+   "hu.dwim.serializer"
+   "hu.dwim.quasi-quote"
+   ;; Tries to write in $HOME
```
I fixed the build of `math`, see commit [653ba45](https://github.com/NixOS/nixpkgs/pull/218870/commits/653ba458347e33ce31fce1c7402678bcdb5bed87)

Uthar replied:
`ubiquitous` is weird: It's creating directories on compile-time. But in this case, the code is much more complex, so I don't know how to safely change it.

Uthar committed:
lisp-modules: use wrapLisp to wrap Lisps

The previous approach of manually repeating a complex pattern inside Common Lisp
implementation package declarations was fragile and hard to change. After
reading python and lua modules code in Nixpkgs, I was able to come up with
something better.

The function `wrapLisp` doesn't need to be inside package declarations so all
the code for wrapping Lisps can be in `all-packages.nix`.

This works by wrapping the `override` function created from `mkDerivation` to
accept a new argument `packageOverrides`.

One problem with this is that `override.__functionArgs` disappears. But one can
look at the source code of a package to discover what can be overridden.

Uthar committed:
clpm: fix eval

kgalkowski-comscore committed:
lisp-modules: second version of wrapLisp

The previous approach of trying to make both the `override` mechanism from
`mkDerivation` and the `overrideScope'` mechanism from `newScope` work together
resulted in hard to understand code, and there was a bug where once overridden
packages would lose the changes on next override with `packageOverrides`.

It's not ideal still, because Lisps created by `mkDerivation` will lose their
`pkgs` after using `override`.

Uthar committed:
lisp-modules: wrap cmucl_binary, gcl and mkcl

(GCL is broken. It doesn't cleanly compile ASDF. Also, of right now, segfaults on errors)

Uthar committed:
clasp-common-lisp: fix eval

Forgot this change in 84eea85ad99ac2dd4c30d22dc32004888ea73b03

Uthar committed:
merge master

Uthar commented:
I should add self-links in lisp.section.md

marsam commented on pkgs/development/lisp-modules/nix-cl.nix:
```diff
@@ -0,0 +1,311 @@
+      pkg,
+
+      # Name of the Lisp exectable
+      program ? pkg.pname,
```
```suggestion
      program ? pkg.meta.mainProgram or pkg.pname,
```

marsam commented on pkgs/development/lisp-modules/nix-cl.nix:
```diff
@@ -0,0 +1,311 @@
+  wrapLisp = {
+    pkg
+    , faslExt
+    , program ? pkg.pname
```
```suggestion
    , program ? pkg.meta.mainProgram or pkg.pname
```

Uthar committed:
lisp-modules: fix typos

Uthar committed:
lisp-modules: use meta.mainProgram before pname

Uthar committed:
lisp-modules: manual: add self links

7c6f434c commented on pkgs/development/compilers/abcl/default.nix:
```diff
@@ -1,4 +1,4 @@
-{lib, stdenv, fetchurl, ant, jre, jdk}:
+{lib, stdenv, fetchurl, ant, jre, jdk, ...}:
```
Is this `...` useful for anything or just a leftover?

Uthar replied:
leftover - rolled it back [c2d2e82](https://github.com/NixOS/nixpkgs/pull/218870/commits/c2d2e8272acad408deb0c66a7b03d67b9f602dfe)

Uthar committed:
lisp-modules: remove left over "..." in argument lists

This was left over after 84eea85ad99ac2dd4c30d22dc32004888ea73b03

Uthar committed:
clasp-common-lisp: fix eval

7c6f434c commented:
Hmm, I wonder if it is a good idea to soften the breaking changes. I.e. _move_ lispPackages to lispPackages-obsolete with a comment to remove in the next release. And maybe even have lispPackages_new-obsolete-layout? So that people can quickly get something-fixed and then migrate over a few months…

Uthar commented:
How about adding a warning like this?

```diff
--- a/pkgs/development/lisp-modules-new/lisp-packages.nix
+++ b/pkgs/development/lisp-modules-new/lisp-packages.nix
@@ -85,7 +85,8 @@ let
   #
   # Wrapper around stdenv.mkDerivation for building ASDF systems.
   #
-  build-asdf-system = makeOverridableLispPackage (
+  build-asdf-system = lib.warn "lispPackages_new is deprecated and will be removed in release 23.05"
+    (makeOverridableLispPackage (
     { pname,
       version,
       src ? null,
@@ -256,7 +257,7 @@ let
         (builtins.concatLists
           (lib.catAttrs "propagatedBuildInputs"
             (builtins.concatLists [[args] lispLibs nativeLibs javaLibs])));
-    })));
+    }))));
 
   # Build the set of lisp packages using `lisp`
   # These packages are defined manually for one reason or another:
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -24220,8 +24220,9 @@ with pkgs;
     inherit clwrapper;
   };
 
-  lispPackages = recurseIntoAttrs (quicklispPackages //
-    (lispPackagesFor (wrapLisp sbcl)));
+  lispPackages = builtins.mapAttrs
+    (n: v: lib.warn "lispPackages is deprecated and will be removed in release 23.05" v)
+    (recurseIntoAttrs (quicklispPackages // (lispPackagesFor (wrapLisp sbcl))));
 
   quicklispPackagesFor = clwrapper: callPackage ../development/lisp-modules/quicklisp-to-nix.nix {
     inherit clwrapper;

```

7c6f434c commented:
Reachable warnings in Nixpkgs? Nah, forbidden

Uthar commented:
Ah, ok (-:

Hmm, moving the current attributes will still break setups. I was thinking, maybe it's better to include a deprecation notice in the next release notes, plus a comment in code, without renaming? Later, remove the current attributes in the second next release, and only then add the new version (so park this pull request until two next releases).

7c6f434c commented:
I think instant-replacement is not too nice either way

Uthar commented:
Ok, so let's sum up:

1. rename the two current `lisp-modules` and `lisp-modules-new` directories
2. rename their attributes with `obsolete` to make it visible that they are old
3. add the new thing

What about the warning using `lib.warn`?

Uthar commented:
Also, `stumpwm` and `nyxt` - should they use the old stuff for now, or come from the new `sbclPackages` ?

Uthar commented:
Perhaps something like the last 2 commits?

7c6f434c commented:
I think `lib.warn` for things reachable by attribute name is, even when not breaking the checks outright, discouraged. The renaming + source code comments about dropping soon should be enough.

A release note entry is welcome but not absolutely indispensable, I think.

If Nyxt and StumpWM work with new stuff, let's roll forward. CLPM, maybe leave back in a working state, and hopefully when it's next updated, it's migrated. Or maybe it gets into Quicklisp…

7c6f434c commented:
Hm wait, you want to say that we can _just_ add this with no renaming (but add a lot of deprecation comments, and a deprecation comment in release notes)? Hmmm. Tempting.

Uthar commented:
> Hm wait, you want to say that we can just add this with no renaming (but add a lot of deprecation comments, and a deprecation comment in release notes)? Hmmm. Tempting.

Yes it is possible, because the only new attribute is `sbclPackages`, otherwise the existing CL implementation attributes ( `sbcl`, ... ) are enhanced. (well there is the `wrapLisp` function, which conflicts with the current one, and is documented in the manual section. But we can rename the old one.)

7c6f434c commented:
My imagination: _rename_ _all_ the attributes to «-obsolete» versions, add _aliases_ (forbidden to use from inside Nixpkgs; maybe we can add lib.warn _there_?), add release notes item about incompatible-but-better interface and migration period.

This is probably over-engineering, but I cannot convince myself of that fact without external help.

Uthar commented:
Oh right, so aliases.nix is a place where old stuff goes. I didn't know that

Uthar committed:
clpm: fix build

It wasn't building because of a too new version of SBCL

Uthar committed:
merge master

7c6f434c commented:
> Such as lispPackages = lib.warn "lispPackages was replaced" lispPackages_obsolete in aliases.nix. This would evolve into lispPackages = throw "lispPackages was removed" in the next release.

Sounds good!

> Are there any features of the current implementation that the new one doesn't support?

I do use esoteric weirdness where you can extract the knowledge how `--eval`/`--load` arguments differ between the implementations, it shouldn't take _long_ to fix, but nominally it is a feature difference.

Uthar commented:
> I do use esoteric weirdness where you can extract the knowledge how --eval/--load arguments differ between the implementations, it shouldn't take long to fix, but nominally it is a feature difference.

I was using --eval/--load flags before, but now switched to passing Lisp code on stdin - primarily this was used to compile and load ASDF

Uthar committed:
lisp-modules: add back the two current implementations

This is to enable a smooth migration to the new one.

Uthar committed:
lisp-modules: deprecate the two current implementations in comments

Uthar commented:
How to unblock ofborg-eval?

Uthar committed:
lisp-modules: add release note about the new manual and the interface

Uthar commented:
does this block merge? https://gist.github.com/GrahamcOfBorg/60be9656aa1258286c965f080d52612f

SuperSandro2000 commented:
> does this block merge? [gist.github.com/GrahamcOfBorg/60be9656aa1258286c965f080d52612f](https://gist.github.com/GrahamcOfBorg/60be9656aa1258286c965f080d52612f)

Yes, it completely blocks this PR.

7c6f434c commented:
So we can put throws but not warnings in aliases? You know what, I am fine with throws…

(They say how to fix them quickly, after all)

Uthar commented:
But adding throws would make the current stuff unusable already. How about keeping the old stuff as it is? Only keeping the deprecation notice in release notes, and then adding the throws in 23.11 ?

7c6f434c commented:
>But adding throws would make the current stuff unusable already. How about keeping the old stuff as it is? Only keeping the deprecation notice in release notes, and then adding the throws in 23.11 ?

23.05: deprecate in Release notes and code comments
23.11: throw and rename for those inattentive
24.05: truly drop

? 

Even better in my opinion…

Uthar commented:
Sounds good

7c6f434c commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -352,6 +352,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `boot.initrd.luks.device.<name>` has a new `tryEmptyPassphrase` option, this is useful for OEM's who need to install an encrypted disk with a future settable passphrase
 
+- Lisp gained a [manual section](https://nixos.org/manual/nixpkgs/stable/#lisp), documenting a new and backwards incompatible interface. The previous interface will not be removed until a future release.
```
I would word assertively…

```suggestion
- Lisp gained a [manual section](https://nixos.org/manual/nixpkgs/stable/#lisp), documenting a new and backwards incompatible interface. The previous interface will be removed in a future release.
```

lukego commented:
woohoo!

dotlambda commented:
This broke eval:
```
       error: attribute 'mysql-client' missing

       at pkgs/development/lisp-modules/ql.nix:59:22:

           58|     cl-mysql = super.cl-mysql.overrideLispAttrs (o: {
           59|       nativeLibs = [ pkgs.mysql-client ];
             |                      ^
           60|     });

```

7c6f434c commented:
Broke eval in a way that OfBorg doesn't catch? Ouch.

dotlambda commented:
> Broke eval in a way that OfBorg doesn't catch? Ouch.

Ofborg currently doesn't check whether aliases are used.

vcunat commented:
For reference, this PR [adds](https://hydra.nixos.org/eval/1793003#tabs-new) over 4k builds of `sbclPackages.*` for `x86_64-linux`.

7c6f434c commented:
(which, all together, are smaller than _one_ big build)

7c6f434c commented:
I would just re-`wrapLisp`, why doesn't it work?

Uthar commented:
Yeah, just re-wrap, until someone figures out how to "keep" the attributes between overrides

> Meta: What's the right place to ask questions about the nix-cl code now that it's upstream?

I would say nixpkgs GitHub issues... or maybe Discourse/Matrix

lukego commented:
I spent half an hour and couldn't get it working with `wrapLisp`. Got into a mess with trying to re-wrap the wrapped package, or not finding a way to unwrap the old wrapper out of the way, and not getting the desired results.

Maybe I'm missing something obvious?

Uthar commented:
Try like this:

```
kpg@e6330:~/scratch/nixpkgs (lisp2) 
$ nix repl
Welcome to Nix 2.11.1. Type :? for help.

nix-repl> :l .
Added 18655 variables.

nix-repl> sbcl2 = sbcl.overrideAttrs (oa: { version = "6.6.6"; }) 

nix-repl> sbcl3 = wrapLisp { pkg = sbcl2; faslExt = "fasl"; }     

nix-repl> sbcl3.withPackages (ps: [ ps.alexandria ])
«derivation /nix/store/mpvhd9cbrmg903hgk8ba29ph97i5lip5-sbcl-with-packages.drv»
```

lukego commented:
> Try like this:

Thanks for the working example! I had goofed.

7c6f434c commented:
Maybe it is a good question if we want to set `hydraPlatforms` to `[]` outside a known-used set (similar to the set of stuff in the old `lispPackages` I guess). As far as I understand, dependencies of things built on Hydra will be available as paths in binary cache even though they won't have job names, so no need to bother too much with tracking closures there.

vcunat commented:
They will be available, though in case of some failures it's easier to deal with them for builds that (also) have their own jobs.

RaitoBezarius commented:
Hey @Uthar ; thank you for this change, I am not very familiar with the maintainers of the Lisp ecosystem, could I invite you to create a NixOS team @NixOS/lisp so that release engineers (or any contributor) can ping you if needed?

SuperSandro2000 commented:
> (which, all together, are smaller than _one_ big build)

It still adds overhead to the hydra evaluator where build time of the derivations does not matter, only count and complexity to evaluate them.

vcunat commented:
Also space.  But even cheap builds if you have lots of them (say, over staging* rebuilds).  The overall farm performance doesn't seem to be capable to exceed around 2 builds per second average, so that's at least half an hour of time on every rebuild.

vcunat commented:
> I also noticed that `clasp-common-lisp` [won't build on arm](https://hydra.nixos.org/build/214768828/log). How to disable it, so that it runs only on x86_64 ?

Ensure correct `meta.platforms` or `meta.broken`, ideally.

7c6f434c commented:
>> Maybe it is a good question if we want to set hydraPlatforms to [] outside a known-used set
>
>Sounds good, I can take this task. I'll create a patch that sets an empty `hydraPlatforms` based on an allow-list of packages, starting from what was available in `lispPackages`

Thanks

>Should we also remove `recurseIntoAttrs` from `lispPackages` and `lispPackages_new`, now that `sbclPackages` is the preferred option?

Indeed!

>I also noticed that `clasp-common-lisp` won't build on arm. How to disable it, so that it runs only on x86_64 ?

Hmm. https://github.com/clasp-developers/clasp/blob/main/src/core/corePackage.cc#L1294 says no upstream support, so just set the platforms, I think.

>> could I invite you to create a NixOS team @NixOS/lisp
>
>Sounds good, created the org and added @lukego @7c6f434c (Hope that's fine with you). 

Fine with me. Maybe also @hraban ? Anyone else interested?

RaitoBezarius commented:
If I may also please ask you to send a PR in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix to add your team and perhaps add some way to insert `meta.maintainers = teams.lisp;` to all lisp modules or packages which you think are in your scope, that would make it extremely awesome.

a12l commented:
How would I go about overriding the packages in 22.11 to enable installation of StumpWM? Unsure how to override a whole ecosystem?

7c6f434c commented:
I think you can just import the package from another checkout
